### PR TITLE
Warn when user sets a property of an inline resource to itself.

### PR DESCRIPTION
**Pull request by jkeiser**
(User will expect "x x" to grab the parent property.)

### CODE REVIEW

**Comment by lamont-granquist** [comment]
Yeah so this still doesn't address #4006 if I read this correctly, but we'll still need this one.  This would close #4250 

:+1:

**Comment by lamont-granquist** [comment]
@chef/client-core needs another review :+1:

**Comment by coderanger** [comment]
I don't really understand what case this is catching.

**Comment by coderanger** [comment]
Ahh I think I see, this is the `self.x` vs enclosing provider's `new_resource.x` issue?

File: lib/chef/property.rb [hunk 1]
```diff
@@ -292,6 +292,15 @@ def get(resource)
         value
 
       else
+        # If we are still compiling this resource (rather than running an action
+        # on it), reading this property before it's been set is likely the wrong
+        # thing. Warn if it's a duplicate of the enclosing provider.
+        if resource.respond_to?(:enclosing_provider) && resource.enclosing_provider && !resource.currently_running_action && !name_property?
+          if resource.enclosing_provider.respond_to?(name)
```
Comment (coderanger) [inline lib/chef/property.rb]: Second nested condition could go on the `&&` chain above.

File: lib/chef/resource.rb [hunk 2]
```diff
@@ -608,15 +609,18 @@ def run_action(action, notification_type=nil, notifying_resource=nil)
...
+    ensure
+      @currently_running_action = nil
+      @elapsed_time = Time.now - start_time
+      # Reporting endpoint doesn't accept a negative resource duration so set it to 0.
+      # A negative value can occur when a resource changes the system time backwards
+      @elapsed_time = 0 if @elapsed_time < 0
+      events.resource_completed(self)
     end
 
+    attr_reader :currently_running_action
```
Comment (coderanger) [inline lib/chef/resource.rb]: Docstring comment plz.

File: lib/chef/log.rb [hunk 3]
```diff
@@ -37,6 +37,13 @@ def self.show_time=(*args)
       end
     end
 
+    def self.caller_location
```
Comment (coderanger) [inline lib/chef/log.rb]: Should have a docstring comment or be marked private.

**Comment by jkeiser** [comment]
@coderanger yep! That's the one.

I could actually fix it so it does what 99% of users expect ... but I think it's worth thinking *that* decision through some more before implementing.

**Comment by coderanger** [comment]
@jkeiser I would probably put a much more descriptive comment about what that warning condition check is doing, wasn't obvious at first glance and it looks like something likely to break in the future.

**Pull request merged**